File: cloud/android/kern/src/main/kotlin/de/innobytix/ahpt/kern/Noise.kt
/*
 * Noise.kt -- Noise IK fuer Android, dritte Umsetzung
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 * Copyright (C) 2026 Manuel Person, InnoBytix-IT
 *
 * WOZU
 * ----
 * Das Gegenstueck zu krypto.py und noise.js, damit die App dieselbe Sprache
 * spricht wie der Kommandozeilen-Client und das Portal. Sie ist die DRITTE
 * Umsetzung desselben Verfahrens -- und wie die beiden anderen wird sie gegen
 * tests/ik_vektoren.json nachgerechnet, Byte fuer Byte. Eine Kryptografie,
 * die nur mit sich selbst funktioniert, faellt genau dabei auf.
 *
 * WARUM BEIDE SUITEN, ANDERS ALS IM PORTAL
 * -----------------------------------------
 * noise.js kann nur AES-GCM, weil WebCrypto kein ChaCha20-Poly1305 kennt.
 * Diese Einschraenkung gibt es hier nicht, also stehen beide zur Verfuegung:
 *
 *   Noise_IK_25519_ChaChaPoly_SHA256   wie ahpt_client.py
 *   Noise_IK_25519_AESGCM_SHA256       wie das Portal
 *
 * Die App spricht ChaChaPoly, also dieselbe wie der Kommandozeilen-Client.
 * AES-GCM ist trotzdem umgesetzt, und zwar nicht auf Vorrat: Nur so lassen
 * sich BEIDE Testvektoren nachrechnen, und ein Fehler in der gemeinsamen
 * Grundrechnung faellt dann zweimal auf statt einmal.
 *
 * WARUM DIE UNTERE SCHICHT VON BOUNCYCASTLE
 * ------------------------------------------
 * Android bringt selbst eine alte, beschnittene BouncyCastle-Fassung mit. Wer
 * einen zweiten Provider registriert, bekommt ein Verhalten, das vom Geraet
 * abhaengt -- und Fehler, die sich auf einem Geraet nicht zeigen.
 * org.bouncycastle.crypto.* und .math.ec.* melden sich nirgends an; sie
 * rechnen nur. Damit ist das Ergebnis auf jedem Geraet dasselbe.
 *
 * DIE ENGLISCHEN NAMEN
 * --------------------
 * MixKey, MixHash, EncryptAndHash, Split heissen wie in der Spezifikation,
 * damit man Zeile fuer Zeile mit ihr und mit den beiden anderen Umsetzungen
 * vergleichen kann. Uebersetzte Namen waeren hier keine Sorgfalt, sondern
 * eine zusaetzliche Fehlerquelle.
 */
package de.innobytix.ahpt.kern

import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.engines.AESEngine
import org.bouncycastle.crypto.macs.HMac
import org.bouncycastle.crypto.modes.ChaCha20Poly1305
import org.bouncycastle.crypto.modes.GCMBlockCipher
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.math.ec.rfc7748.X25519
import java.security.SecureRandom

/** Bindet den Handshake an DIESES Protokoll -- muss beim Agenten gleich sein. */
val AHPT_PROLOG: ByteArray = "AHPT-Privat/1".toByteArray(Charsets.UTF_8)

class NoiseFehler(meldung: String) : Exception(meldung)

/* ------------------------------------------------------------- Werkzeug */

fun hexZuBytes(h: String): ByteArray {
    val s = h.trim().lowercase().filter { it in "0123456789abcdef" }
    if (s.length % 2 != 0) throw NoiseFehler("Hexwert hat ungerade Laenge")
    return ByteArray(s.length / 2) {
        ((Character.digit(s[it * 2], 16) shl 4) or
                Character.digit(s[it * 2 + 1], 16)).toByte()
    }
}

fun bytesZuHex(b: ByteArray): String = b.joinToString("") { "%02x".format(it) }

fun verkette(vararg teile: ByteArray): ByteArray {
    val r = ByteArray(teile.sumOf { it.size })
    var i = 0
    for (t in teile) {
        t.copyInto(r, i)
        i += t.size
    }
    return r
}

/**
 * Vergleich in gleichbleibender Zeit.
 *
 * Ein Vergleich, der beim ersten Unterschied abbricht, verraet ueber seine
 * Laufzeit, WIE WEIT zwei Werte uebereinstimmen. Wer raten darf und die Zeit
 * messen kann, raet damit Stelle fuer Stelle statt auf einen Schlag.
 */
fun gleich(a: ByteArray, b: ByteArray): Boolean {
    if (a.size != b.size) return false
    var d = 0
    for (i in a.indices) d = d or (a[i].toInt() xor b[i].toInt())
    return d == 0
}

/* -------------------------------------------------------- Grundrechnung */

fun sha256(daten: ByteArray): ByteArray {
    val d = SHA256Digest()
    d.update(daten, 0, daten.size)
    return ByteArray(d.digestSize).also { d.doFinal(it, 0) }
}

fun hmacSha256(schluessel: ByteArray, daten: ByteArray): ByteArray {
    val m = HMac(SHA256Digest())
    m.init(KeyParameter(schluessel))
    m.update(daten, 0, daten.size)
    return ByteArray(m.macSize).also { m.doFinal(it, 0) }
}

/**
 * Die HKDF-Variante der NOISE-Spezifikation.
 *
 * ACHTUNG: Das ist NICHT die uebliche HKDF. Noise verkettet HMAC-Aufrufe in
 * einer eigenen Form -- jeder Ausgang geht in den naechsten ein --, und es
 * gibt kein `info`-Feld. Wer hier eine fertige HKDF einsetzt, bekommt andere
 * Schluessel und merkt es erst, wenn die Testvektoren nicht mehr stimmen.
 * Genau dafuer gibt es sie.
 */
fun noiseHkdf(ck: ByteArray, ikm: ByteArray, anzahl: Int): List<ByteArray> {
    val temp = hmacSha256(ck, ikm)
    val o1 = hmacSha256(temp, byteArrayOf(1))
    if (anzahl == 1) return listOf(o1)
    val o2 = hmacSha256(temp, verkette(o1, byteArrayOf(2)))
    if (anzahl == 2) return listOf(o1, o2)
    val o3 = hmacSha256(temp, verkette(o2, byteArrayOf(3)))
    return listOf(o1, o2, o3)
}

/* --------------------------------------------------------------- X25519 */

private val zufall = SecureRandom()

object Dh {
    /** Oeffentlicher Teil zu einem privaten Rohschluessel. */
    fun oeffentlichZu(privat32: ByteArray): ByteArray {
        pruefe32(privat32, "privater Schluessel")
        return ByteArray(32).also { X25519.scalarMultBase(privat32, 0, it, 0) }
    }

    /** Gemeinsames Geheimnis aus eigenem privaten und fremdem oeffentlichen. */
    fun dh(privat32: ByteArray, oeffentlich32: ByteArray): ByteArray {
        pruefe32(privat32, "privater Schluessel")
        pruefe32(oeffentlich32, "oeffentlicher Schluessel")
        val r = ByteArray(32)
        // Den Rueckgabewert zu verwerfen waere falsch: X25519 meldet mit
        // `false` einen Punkt kleiner Ordnung. Das Ergebnis waere dann lauter
        // Nullen, und zwar fuer JEDEN privaten Schluessel -- ein Gegenueber
        // koennte damit ein "gemeinsames Geheimnis" erzwingen, das es vorher
        // kennt.
        if (!X25519.calculateAgreement(privat32, 0, oeffentlich32, 0, r, 0)) {
            throw NoiseFehler("X25519: unbrauchbarer oeffentlicher Schluessel")
        }
        return r
    }

    /** Neues Schluesselpaar: (privat, oeffentlich). */
    fun schluesselpaar(): Pair<ByteArray, ByteArray> {
        val privat = ByteArray(32).also { zufall.nextBytes(it) }
        return privat to oeffentlichZu(privat)
    }

    private fun pruefe32(b: ByteArray, was: String) {
        if (b.size != 32) throw NoiseFehler("$was: 32 Bytes erwartet, nicht ${b.size}")
    }
}

/* ---------------------------------------------------------------- Suite */

enum class Suite(val protokollName: String) {
    ChaChaPoly("Noise_IK_25519_ChaChaPoly_SHA256"),
    AesGcm("Noise_IK_25519_AESGCM_SHA256");

    companion object {
        fun ausName(name: String): Suite =
            entries.firstOrNull { it.protokollName == name }
                ?: throw NoiseFehler("unbekannte Suite: $name")
    }
}

/* ---------------------------------------------------------- CipherState */

class CipherState(private val suite: Suite) {
    private var k: ByteArray? = null
    private var n: Long = 0

    fun InitializeKey(schluessel: ByteArray?) {
        k = schluessel
        n = 0
    }

    fun HasKey(): Boolean = k != null

    /**
     * 96 Bit: 32 Bit Null, dann der Zaehler.
     *
     * AES-GCM verlangt BIG ENDIAN, ChaCha20-Poly1305 LITTLE ENDIAN. Das ist
     * keine Schlamperei der Spezifikation, sondern folgt den jeweiligen
     * Originalarbeiten. Wer beides gleich behandelt, baut etwas, das mit sich
     * selbst funktioniert und mit keiner anderen Umsetzung. Der Fehler faellt
     * erst bei der ZWEITEN Nachricht je Richtung auf, weil null in beiden
     * Reihenfolgen dasselbe ist -- und bis dahin sieht alles gut aus.
     */
    private fun nonce(): ByteArray {
        val iv = ByteArray(12)
        for (i in 0 until 8) {
            val b = ((n ushr (8 * i)) and 0xff).toByte()
            if (suite == Suite.AesGcm) iv[11 - i] = b else iv[4 + i] = b
        }
        return iv
    }

    private fun lauf(verschluesseln: Boolean, ad: ByteArray, ein: ByteArray): ByteArray {
        val schluessel = k ?: return ein
        val p = AEADParameters(KeyParameter(schluessel), 128, nonce(), ad)
        val c = if (suite == Suite.AesGcm)
            GCMBlockCipher.newInstance(AESEngine.newInstance())
        else
            ChaCha20Poly1305()
        c.init(verschluesseln, p)
        val aus = ByteArray(c.getOutputSize(ein.size))
        val geschrieben = c.processBytes(ein, 0, ein.size, aus, 0)
        c.doFinal(aus, geschrieben)
        return aus
    }

    fun EncryptWithAd(ad: ByteArray, klartext: ByteArray): ByteArray {
        if (k == null) return klartext
        return lauf(true, ad, klartext).also { n++ }
    }

    fun DecryptWithAd(ad: ByteArray, geheim: ByteArray): ByteArray {
        if (k == null) return geheim
        val p = try {
            lauf(false, ad, geheim)
        } catch (e: Exception) {
            // Absichtlich wortkarg: Eine Meldung, die zwischen "falscher
            // Schluessel" und "veraendert" unterscheidet, ist ein
            // Auskunftsdienst fuer den, der es versucht.
            throw NoiseFehler("Entschluesselung fehlgeschlagen")
        }
        n++
        return p
    }
}

/* ------------------------------------------------------- SymmetricState */

class SymmetricState(private val suite: Suite) {
    var h: ByteArray
        private set
    private var ck: ByteArray
    private val cs = CipherState(suite)

    init {
        val name = suite.protokollName.toByteArray(Charsets.UTF_8)
        // Ist der Name hoechstens so lang wie der Hash, wird er mit Nullen
        // aufgefuellt statt gehasht -- so steht es in der Spezifikation.
        h = if (name.size <= 32) ByteArray(32).also { name.copyInto(it) } else sha256(name)
        ck = h.copyOf()
    }

    fun MixKey(ikm: ByteArray) {
        val (neuCk, tempK) = noiseHkdf(ck, ikm, 2)
        ck = neuCk
        cs.InitializeKey(tempK)
    }

    fun MixHash(daten: ByteArray) {
        h = sha256(verkette(h, daten))
    }

    fun EncryptAndHash(klartext: ByteArray): ByteArray =
        cs.EncryptWithAd(h, klartext).also { MixHash(it) }

    fun DecryptAndHash(geheim: ByteArray): ByteArray {
        // Gehasht wird der GEHEIMTEXT, nicht der Klartext -- sonst haetten
        // beide Seiten verschiedene Hashes, sobald ein Feld unverschluesselt
        // bleibt.
        val p = cs.DecryptWithAd(h, geheim)
        MixHash(geheim)
        return p
    }

    fun Split(): Pair<CipherState, CipherState> {
        val (t1, t2) = noiseHkdf(ck, ByteArray(0), 2)
        return CipherState(suite).also { it.InitializeKey(t1) } to
                CipherState(suite).also { it.InitializeKey(t2) }
    }
}

/* ----------------------------------------------------------- HandshakeIK
 *
 *   IK:
 *     <- s
 *     ...
 *     -> e, es, s, ss
 *     <- e, ee, se
 *
 * Die App ist immer der Initiator: Sie kennt den statischen Schluessel des
 * Agenten, weil du ihn von Hand eingetragen hast. Die Gegenrichtung
 * (Responder) ist deshalb NICHT umgesetzt -- sie liefe hier nie, und
 * unbenutzter Code an einer Sicherheitsschranke wird auch nicht mitgeprueft.
 */
class HandshakeIK(
    suite: Suite,
    prologue: ByteArray,
    /** Eigener privater Schluessel, 32 rohe Bytes. */
    private val sRoh: ByteArray,
    /** Oeffentlicher Schluessel des Agenten, 32 rohe Bytes. */
    private val rs: ByteArray,
    /** Fluechtiger Schluessel -- nur fuer die Testvektoren vorzugeben. */
    eRoh: ByteArray? = null,
) {
    private val ss = SymmetricState(suite)
    private val sPub = Dh.oeffentlichZu(sRoh)
    private var e: ByteArray? = eRoh
    private var fertig = false

    init {
        ss.MixHash(prologue)
        // Vornachricht "<- s": Der statische Schluessel des Antwortenden ist
        // dem Initiator schon bekannt und geht in den Hash ein.
        ss.MixHash(rs)
    }

    fun schreibeNachricht1(nutzlast: ByteArray): ByteArray {
        val eigenesE = e ?: Dh.schluesselpaar().first.also { e = it }
        val ePub = Dh.oeffentlichZu(eigenesE)
        ss.MixHash(ePub)                                   // e
        ss.MixKey(Dh.dh(eigenesE, rs))                     // es
        val s = ss.EncryptAndHash(sPub)                    // s
        ss.MixKey(Dh.dh(sRoh, rs))                         // ss
        return verkette(ePub, s, ss.EncryptAndHash(nutzlast))
    }

    fun liesNachricht2(nachricht: ByteArray): ByteArray {
        if (nachricht.size < 32 + 16) throw NoiseFehler("Nachricht 2 zu kurz")
        val eigenesE = e ?: throw NoiseFehler("Nachricht 1 wurde nie geschrieben")
        val re = nachricht.copyOfRange(0, 32)
        ss.MixHash(re)                                     // e
        ss.MixKey(Dh.dh(eigenesE, re))                     // ee
        ss.MixKey(Dh.dh(sRoh, re))                         // se
        val klartext = ss.DecryptAndHash(nachricht.copyOfRange(32, nachricht.size))
        fertig = true
        return klartext
    }

    fun handshakeHash(): ByteArray = ss.h

    /** Erst nach dem Handshake: die beiden Transportschluessel. */
    fun transportschluessel(): Pair<CipherState, CipherState> {
        if (!fertig) throw NoiseFehler("Handshake noch nicht abgeschlossen")
        return ss.Split()
    }
}

File: cloud3000/portal/noise.js
/*
 * noise.js -- Noise IK im Browser, nur mit WebCrypto
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 * Copyright (C) 2026 Manuel Person, InnoBytix-IT
 *
 * WOZU
 * ----
 * Das Gegenstueck zu krypto.py, damit das Portal dieselbe Sprache spricht
 * wie der Kommandozeilen-Client. Ohne Bibliothek, ohne Bauschritt -- eine
 * Datei, die der Browser direkt laedt.
 *
 * WARUM AES-GCM UND NICHT ChaCha20-Poly1305
 * ------------------------------------------
 * WebCrypto kennt kein ChaCha20-Poly1305. Es gaebe zwei Auswege: eine
 * WASM-Bibliothek einbinden (eine Abhaengigkeit, die man pruefen muesste)
 * oder die andere Suite nehmen, die WebCrypto beherrscht.
 *
 * Hier steht die zweite: `Noise_IK_25519_AESGCM_SHA256`. Das ist eine
 * regulaere Noise-Suite, gegen dieselben offiziellen Testvektoren geprueft
 * wie die ChaCha-Variante -- siehe `pruefe_noise.html`, die genau das im
 * Browser nachrechnet.
 *
 * Der Agent nimmt beide an und antwortet in derselben, die er bekommen hat.
 *
 * WAS DIESE DATEI NICHT LEISTET -- BITTE LESEN
 * ---------------------------------------------
 * Sie kommt vom Webspace. Der Hoster koennte sie austauschen und damit auch
 * den Schluessel, gegen den verschluesselt wird. Gegen einen BOESWILLIGEN
 * HOSTER schuetzt ein Portal auf seinem eigenen Server also nicht -- das
 * kann es nicht, egal wie gut die Kryptografie ist.
 *
 * Wogegen es sehr wohl schuetzt, und das ist der eigentliche Zweck: gegen
 * JEDEN ANDEREN. Ohne Verschluesselung kann jeder im Internet die
 * Warteschlange abrufen, die Marken nehmen und Fragen wie Antworten
 * mitlesen. Genau das ist damit zu.
 *
 * Wer auch dem Hoster nicht traut, nimmt den Kommandozeilen-Client oder
 * legt diese Seite als Datei auf den eigenen Rechner. Dann ist der Webspace
 * nur noch Briefkasten.
 *
 * DIE ENGLISCHEN NAMEN
 * --------------------
 * MixKey, MixHash, EncryptAndHash, Split heissen wie in der Spezifikation,
 * damit man Zeile fuer Zeile vergleichen kann. Uebersetzte Namen waeren
 * hier keine Sorgfalt, sondern eine zusaetzliche Fehlerquelle.
 */

'use strict';

const AHPT_PROLOG = new TextEncoder().encode('AHPT-Privat/1');

/* ---------------------------------------------------------- Werkzeug */

function hexZuBytes(h) {
  h = (h || '').trim().toLowerCase().replace(/[^0-9a-f]/g, '');
  if (h.length % 2) throw new Error('Hexwert hat ungerade Laenge');
  const b = new Uint8Array(h.length / 2);
  for (let i = 0; i < b.length; i++) b[i] = parseInt(h.substr(i * 2, 2), 16);
  return b;
}

function bytesZuHex(b) {
  return Array.from(b).map(x => x.toString(16).padStart(2, '0')).join('');
}

function bytesZuB64(b) {
  let s = '';
  const CH = 0x8000;                       // in Haeppchen, sonst Stapelfehler
  for (let i = 0; i < b.length; i += CH) {
    s += String.fromCharCode.apply(null, b.subarray(i, i + CH));
  }
  return btoa(s);
}

function b64ZuBytes(s) {
  const roh = atob(s);
  const b = new Uint8Array(roh.length);
  for (let i = 0; i < roh.length; i++) b[i] = roh.charCodeAt(i);
  return b;
}

function verkette(...teile) {
  let n = 0;
  for (const t of teile) n += t.length;
  const r = new Uint8Array(n);
  let i = 0;
  for (const t of teile) { r.set(t, i); i += t.length; }
  return r;
}

function gleich(a, b) {
  if (a.length !== b.length) return false;
  let d = 0;
  for (let i = 0; i < a.length; i++) d |= a[i] ^ b[i];
  return d === 0;
}

/* ------------------------------------------------------- Grundrechnung */

async function sha256(daten) {
  return new Uint8Array(await crypto.subtle.digest('SHA-256', daten));
}

async function hmacSha256(schluessel, daten) {
  const k = await crypto.subtle.importKey(
    'raw', schluessel, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']);
  return new Uint8Array(await crypto.subtle.sign('HMAC', k, daten));
}

/*
 * Die HKDF-Variante der NOISE-Spezifikation.
 *
 * ACHTUNG: Das ist NICHT WebCryptos HKDF. Noise verkettet HMAC-Aufrufe in
 * einer eigenen Form (jeder Ausgang geht in den naechsten ein), ohne
 * `info`-Feld. Wer hier die eingebaute HKDF benutzt, bekommt andere
 * Schluessel -- und merkt es erst, wenn die Testvektoren nicht mehr stimmen.
 * Genau dafuer gibt es sie.
 */
async function noiseHkdf(ck, ikm, anzahl) {
  const temp = await hmacSha256(ck, ikm);
  const o1 = await hmacSha256(temp, new Uint8Array([1]));
  if (anzahl === 1) return [o1];
  const o2 = await hmacSha256(temp, verkette(o1, new Uint8Array([2])));
  if (anzahl === 2) return [o1, o2];
  const o3 = await hmacSha256(temp, verkette(o2, new Uint8Array([3])));
  return [o1, o2, o3];
}

/* ------------------------------------------------------------- X25519 */

// WebCrypto nimmt einen privaten X25519-Schluessel nicht als 32 rohe Bytes
// entgegen -- nur als PKCS8 oder JWK. Diese 16 Bytes sind der feste
// PKCS8-Rahmen fuer X25519; danach folgt der rohe Schluessel.
const PKCS8_X25519 = new Uint8Array([
  0x30, 0x2e, 0x02, 0x01, 0x00, 0x30, 0x05, 0x06,
  0x03, 0x2b, 0x65, 0x6e, 0x04, 0x22, 0x04, 0x20]);

async function privatImportieren(rohe32) {
  if (rohe32.length !== 32) throw new Error('privater Schluessel: 32 Bytes erwartet');
  return crypto.subtle.importKey(
    'pkcs8', verkette(PKCS8_X25519, rohe32),
    { name: 'X25519' }, false, ['deriveBits']);
}

async function oeffentlichImportieren(rohe32) {
  if (rohe32.length !== 32) throw new Error('oeffentlicher Schluessel: 32 Bytes erwartet');
  return crypto.subtle.importKey('raw', rohe32, { name: 'X25519' }, true, []);
}

async function dh(privatKey, oeffentlichRoh) {
  const pub = await oeffentlichImportieren(oeffentlichRoh);
  return new Uint8Array(await crypto.subtle.deriveBits(
    { name: 'X25519', public: pub }, privatKey, 256));
}

/** Neues Schluesselpaar. Rueckgabe: {privat: CryptoKey, roh, oeffentlich}. */
async function schluesselpaar() {
  const p = await crypto.subtle.generateKey({ name: 'X25519' }, true,
                                            ['deriveBits']);
  const oeff = new Uint8Array(await crypto.subtle.exportKey('raw', p.publicKey));
  const pkcs8 = new Uint8Array(await crypto.subtle.exportKey('pkcs8', p.privateKey));
  return {
    privat: p.privateKey,
    roh: pkcs8.slice(PKCS8_X25519.length),   // die 32 Bytes am Ende
    oeffentlich: oeff
  };
}

/** Zu einem privaten Rohschluessel den oeffentlichen Teil bestimmen. */
async function oeffentlichZu(rohe32) {
  // WebCrypto kann aus einem PKCS8-Import den oeffentlichen Teil nicht
  // herausgeben. Der Umweg: als JWK importieren geht nur mit bekanntem `x`.
  // Also rechnen wir ihn: X25519 mit dem Basispunkt.
  const BASIS = new Uint8Array(32); BASIS[0] = 9;
  const k = await privatImportieren(rohe32);
  return dh(k, BASIS);
}

/* --------------------------------------------------------- CipherState */

class CipherState {
  constructor() { this.k = null; this.n = 0n; }

  InitializeKey(k) { this.k = k; this.n = 0n; }
  HasKey() { return this.k !== null; }

  // 96 Bit: 32 Bit Null, dann der Zaehler.
  //
  // AES-GCM verlangt BIG ENDIAN -- ChaCha20-Poly1305 verlangt little.
  // Das ist keine Schlamperei der Spezifikation, sondern folgt den
  // jeweiligen Originalarbeiten. Wer beides gleich behandelt, baut etwas,
  // das mit sich selbst funktioniert und mit keiner anderen Umsetzung. Der
  // Fehler faellt erst bei der ZWEITEN Nachricht je Richtung auf, weil
  // null in beiden Reihenfolgen dasselbe ist.
  _nonce() {
    const iv = new Uint8Array(12);
    const dv = new DataView(iv.buffer);
    dv.setBigUint64(4, this.n, false);      // false = big endian
    return iv;
  }

  async _key() {
    return crypto.subtle.importKey('raw', this.k, { name: 'AES-GCM' },
                                   false, ['encrypt', 'decrypt']);
  }

  async EncryptWithAd(ad, klartext) {
    if (this.k === null) return klartext;
    const c = new Uint8Array(await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv: this._nonce(), additionalData: ad, tagLength: 128 },
      await this._key(), klartext));
    this.n += 1n;
    return c;
  }

  async DecryptWithAd(ad, geheim) {
    if (this.k === null) return geheim;
    let p;
    try {
      p = new Uint8Array(await crypto.subtle.decrypt(
        { name: 'AES-GCM', iv: this._nonce(), additionalData: ad, tagLength: 128 },
        await this._key(), geheim));
    } catch (e) {
      // Absichtlich wortkarg: Eine Meldung, die zwischen "falscher
      // Schluessel" und "veraendert" unterscheidet, ist ein Auskunftsdienst
      // fuer den, der es versucht.
      throw new Error('Entschluesselung fehlgeschlagen');
    }
    this.n += 1n;
    return p;
  }
}

/* ------------------------------------------------------ SymmetricState */

class SymmetricState {
  static async neu(protokollName) {
    const s = new SymmetricState();
    const name = new TextEncoder().encode(protokollName);
    // Ist der Name hoechstens so lang wie der Hash, wird er mit Nullen
    // aufgefuellt statt gehasht. Unserer ist genau 28 Zeichen lang.
    if (name.length <= 32) {
      s.h = new Uint8Array(32);
      s.h.set(name, 0);
    } else {
      s.h = await sha256(name);
    }
    s.ck = s.h.slice();
    s.cs = new CipherState();
    return s;
  }

  async MixKey(ikm) {
    const [ck, tempK] = await noiseHkdf(this.ck, ikm, 2);
    this.ck = ck;
    this.cs.InitializeKey(tempK);
  }

  async MixHash(daten) { this.h = await sha256(verkette(this.h, daten)); }

  async EncryptAndHash(klartext) {
    const c = await this.cs.EncryptWithAd(this.h, klartext);
    await this.MixHash(c);
    return c;
  }

  async DecryptAndHash(geheim) {
    // Gehasht wird der GEHEIMTEXT, nicht der Klartext -- sonst haetten beide
    // Seiten verschiedene Hashes, sobald ein Feld unverschluesselt bleibt.
    const p = await this.cs.DecryptWithAd(this.h, geheim);
    await this.MixHash(geheim);
    return p;
  }

  async Split() {
    const [t1, t2] = await noiseHkdf(this.ck, new Uint8Array(0), 2);
    const c1 = new CipherState(); c1.InitializeKey(t1);
    const c2 = new CipherState(); c2.InitializeKey(t2);
    return [c1, c2];
  }
}

/* ---------------------------------------------------------- HandshakeIK
 *
 *   IK:
 *     <- s
 *     ...
 *     -> e, es, s, ss
 *     <- e, ee, se
 *
 * Der Browser ist immer der Initiator: Er kennt den statischen Schluessel
 * des Agenten, weil du ihn von Hand eingetragen hast.
 */

const PROTOKOLL_NAME = 'Noise_IK_25519_AESGCM_SHA256';

class HandshakeIK {
  /**
   * @param prologue  bindet den Handshake an dieses Protokoll
   * @param sRoh      eigener privater Schluessel, 32 rohe Bytes
   * @param rs        oeffentlicher Schluessel des Agenten, 32 rohe Bytes
   * @param eRoh      fluechtiger Schluessel -- nur fuer Testvektoren
   */
  static async neu(prologue, sRoh, rs, eRoh) {
    const h = new HandshakeIK();
    h.ss = await SymmetricState.neu(PROTOKOLL_NAME);
    await h.ss.MixHash(prologue);
    h.sRoh = sRoh;
    h.s = await privatImportieren(sRoh);
    h.sPub = await oeffentlichZu(sRoh);
    h.rs = rs;
    h.re = null;
    h.fertig = false;
    if (eRoh) {
      h.eRoh = eRoh;
      h.e = await privatImportieren(eRoh);
      h.ePub = await oeffentlichZu(eRoh);
    }
    // Vornachricht "<- s": Der statische Schluessel des Antwortenden ist dem
    // Initiator schon bekannt und geht in den Hash ein.
    await h.ss.MixHash(rs);
    return h;
  }

  async schreibeNachricht1(nutzlast) {
    if (!this.e) {
      const p = await schluesselpaar();
      this.e = p.privat; this.ePub = p.oeffentlich; this.eRoh = p.roh;
    }
    let puffer = this.ePub;
    await this.ss.MixHash(this.ePub);                       // e
    await this.ss.MixKey(await dh(this.e, this.rs));        // es
    puffer = verkette(puffer, await this.ss.EncryptAndHash(this.sPub));  // s
    await this.ss.MixKey(await dh(this.s, this.rs));        // ss
    puffer = verkette(puffer, await this.ss.EncryptAndHash(nutzlast));
    return puffer;
  }

  async liesNachricht2(nachricht) {
    if (nachricht.length < 32 + 16) throw new Error('Nachricht 2 zu kurz');
    this.re = nachricht.slice(0, 32);
    await this.ss.MixHash(this.re);                         // e
    await this.ss.MixKey(await dh(this.e, this.re));        // ee
    await this.ss.MixKey(await dh(this.s, this.re));        // se
    const klartext = await this.ss.DecryptAndHash(nachricht.slice(32));
    this.fertig = true;
    return klartext;
  }

  handshakeHash() { return this.ss.h; }

  async transportschluessel() {
    if (!this.fertig) throw new Error('Handshake noch nicht abgeschlossen');
    return this.ss.Split();
  }
}

/* --------------------------------------------------------- Verfuegbarkeit */

/**
 * Kann dieser Browser, was wir brauchen?
 *
 * X25519 in WebCrypto ist noch nicht ueberall da. Lieber EINMAL klar sagen,
 * dass es nicht geht, als bei jedem Klick eine unverstaendliche Fehlermeldung
 * zu erzeugen.
 */
async function kryptoVerfuegbar() {
  if (typeof window !== 'undefined' && !window.isSecureContext) {
    return { ok: false, grund: 'Die Seite laeuft nicht in einem sicheren ' +
             'Kontext. WebCrypto gibt es nur ueber https oder auf localhost.' };
  }
  if (typeof crypto === 'undefined' || !crypto.subtle) {
    return { ok: false, grund: 'Dieser Browser hat kein WebCrypto.' };
  }
  try {
    await crypto.subtle.generateKey({ name: 'X25519' }, true, ['deriveBits']);
  } catch (e) {
    return { ok: false, grund: 'Dieser Browser kann kein X25519. ' +
             'Noetig sind Chrome 133+, Firefox 132+ oder Safari 17.4+.' };
  }
  return { ok: true };
}

/* --------------------------------------------------------------------
 * Fuer die Pruefung ausserhalb des Browsers.
 *
 * `tests/pruefe_noise_js.cjs` rechnet diese Datei gegen dieselben
 * offiziellen Testvektoren nach wie krypto.py. Ohne diesen Block muesste
 * der Test den Quelltext auswerten -- und dann prueft er eine Kopie, nicht
 * das Original.
 *
 * Im Browser ist `module` nicht definiert; der Block tut dort nichts.
 * -------------------------------------------------------------------- */
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    HandshakeIK, CipherState, SymmetricState, noiseHkdf, schluesselpaar,
    oeffentlichZu, privatImportieren, dh, sha256, hmacSha256,
    hexZuBytes, bytesZuHex, bytesZuB64, b64ZuBytes, verkette, gleich,
    kryptoVerfuegbar, AHPT_PROLOG, PROTOKOLL_NAME
  };
}
